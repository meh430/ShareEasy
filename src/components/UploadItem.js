import React from "react";
import "../App.css";
import { CopyToClipboard } from "react-copy-to-clipboard";

const iconStyle = {
    fontSize: "52px",
    color: "red",
    margin: "14px",
};
const mediaIcons = {
    image: "fas fa-file-image",
    video: "far fa-file-video",
    audio: "far fa-file-audio",
    pdf: "far fa-file-pdf",
    text: "far fa-file-alt",
    zip: "far fa-file-archive",
    upload: "fas fa-upload",
};


//fileName: string
//fileType: string
//downloadLink: string
//expireTime: optional string
export class UploadItem extends React.Component {
    constructor(props) {
        super(props);
        this.displayLink = this.displayLink.bind(this);
        this.displayMediaIcon = this.displayMediaIcon.bind(this);
        this.copiedText = this.copiedText.bind(this);
        this.state = { copyState: "#a5bef0" };
    }

    copiedText(event) {
        this.setState({ copyState: "#c4f5d6", copyMargin: "4px" });
    }

    displayMediaIcon() {
        if (this.props.fileName.length === 0 || this.props.fileType.length === 0) {
            return mediaIcons.upload;
        } else {
            const mediaKeys = Object.keys(mediaIcons);
            for (let i = 0; i < mediaKeys.length; i++) {
                if (this.props.fileType.includes(mediaKeys[i])) {
                    return mediaIcons[mediaKeys[i]];
                }
            }

            return mediaIcons.text;
        }
    }

    displayLink() {
        if (this.props.downloadLink.length !== 0) {
            return (
                <div className="dLink" style={{ backgroundColor: this.state.copyState }}>
                    {this.props.downloadLink}{" "}
                    <CopyToClipboard text={this.props.downloadLink}>
                        <div
                            onClick={this.copiedText}
                            style={{
                                borderLeft: "solid 1px black",
                                display: "inline-block",
                                padding: "8px",
                                paddingLeft: "12px",
                                marginLeft: "12px",
                                fontSize: "16px",
                            }}
                        >
                            <i className="far fa-clone" style={{ cursor: "pointer" }} onClick={this.copiedText}></i>
                        </div>
                    </CopyToClipboard>
                </div>
            );
        }
    }

    render() {
        return (
            <div className="upItem">
                <div className="upItemText">
                    <i className={this.displayMediaIcon().toString()} style={iconStyle} />
                    <h3 className="fileName">
                        {this.props.fileName.length === 0 ? "Please upload a file..." : this.props.fileName}
                    </h3>
                </div>

                {this.displayLink()}

                {this.props.expireTime ? <h3 className="fileName">{"Expires in " + this.props.expireTime + " hrs"}</h3> : <div></div>}
            </div>
        );
    }
}
