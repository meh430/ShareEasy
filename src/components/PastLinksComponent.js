import React from "react";
import "../App.css";
import { UploadItem } from "./UploadItem";

let uploads = JSON.parse(localStorage.getItem("SAVED_LINKS"));
uploads = uploads ? uploads : [];

const createListItems = (uploadList) => {
    let ret = [];
    for (let i = 0; i < uploadList.length; i++) {
        const currLink = uploadList[i]
        ret.push(
            <li style={{display: "inline-block"}}>
                <UploadItem
                    fileType={currLink.fileType}
                    fileName={currLink.fileName}
                    expireTime="24 hrs"
                    downloadLink={currLink.dLink}
                />
            </li>
        );
    }

    return ret.length > 0 ? ret : [<li><h3 className="fileName">No Uploads...</h3></li>];
};

export class PastLinksComponent extends React.Component {
    constructor(props) {
        super(props);
        this.state = { pastLinks: createListItems(uploads) };
        this.updateList = this.updateList.bind(this)
        setInterval(this.updateList, 1000);
        console.log(uploads)
    }

    updateList() {
        uploads = JSON.parse(localStorage.getItem("SAVED_LINKS"));
        uploads = uploads ? uploads : [];
        this.setState({pastLinks: createListItems(uploads)})
    }

    render() {
        return (
            <div className="section">
                <h3 className="sectionInfo">Generated download links</h3>
                <ul className="pastUploads">
                    {this.state.pastLinks}
                </ul>
            </div>
        );
    }
}
