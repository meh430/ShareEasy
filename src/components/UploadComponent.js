import React from "react";
import "../App.css";
import firebase from "../Firebase";
import { UploadItem } from "./UploadItem";
import { shortenReqHeaders } from "../Credentials";
const SAVED_LINKS = 'SAVED_LINKS'
const storageRef = firebase.storage().ref();
let uploadTask;
let uploads = JSON.parse(localStorage.getItem(SAVED_LINKS))
uploads = uploads ? uploads : []

//link item: {fileName, fileType, link}

export class UploadComponent extends React.Component {
    constructor(props) {
        super(props);
        this.uploadClicked = this.uploadClicked.bind(this);
        this.onFileSelected = this.onFileSelected.bind(this);
        this.displayProgress = this.displayProgress.bind(this);
        this.state = { fileName: "", downloadLink: "", fileType: "", uploadRunning: false, progress: 0 };
    }

    displayProgress() {
        if (this.state.uploadRunning) {
            return <progress value={this.state.progress} max="100" style={{ width: "80%" }} />;
        } else {
            return <i style={{ display: "none" }} />;
        }
    }

    uploadClicked(event) {
        if (!this.state.uploadRunning) {
            this.setState({ fileName: "", downloadLink: "", fileType: "", uploadRunning: false, progress: 0 });
            this.refs.fileUpload.click();
        }
    }

    //TODO: handle errors
    onFileSelected(event) {
        event.stopPropagation();
        event.preventDefault();
        const selectedFile = event.target.files[0];
        if (selectedFile && !this.state.uploadRunning) {
            let fName = selectedFile["name"];
            fName = fName.replace(/\s/g, "");
            uploadTask = storageRef.child(fName).put(selectedFile);
            this.setState({ fileName: fName, fileType: selectedFile.type });
            uploadTask.on(
                "state_changed",
                (snapshot) => {
                    const progress = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
                    this.setState({ progress: progress, uploadRunning: true });
                },
                (err) => {
                    console.log(err);
                },
                () => {
                    uploadTask.snapshot.ref.getDownloadURL().then((downloadURL) => {
                        let shortenRequest = {
                            destination: downloadURL,
                        };

                        fetch("https://api.rebrandly.com/v1/links", {
                            method: "POST",
                            headers: shortenReqHeaders,
                            body: JSON.stringify(shortenRequest),
                        })
                            .then((res) => {
                                return res.json();
                            })
                            .then((link) => {
                                console.log(link);
                                this.setState({
                                    downloadLink: link.shortUrl,
                                    uploadRunning: false,
                                    progress: 100,
                                    fileName: fName,
                                    fileType: selectedFile.type,
                                });

                                uploads.push({ fileName: fName, fileType: selectedFile.type, dLink: link.shortUrl });
                                localStorage.setItem(SAVED_LINKS, JSON.stringify(uploads));
                            });
                    });
                }
            );
            console.log(selectedFile);
            this.setState({ fileName: fName, uploadRunning: true });
        }
    }

    render() {
        //TODO: figure out expiration countdown and auto deletion
        return (
            <div className="section">
                <h3 className="sectionInfo">
                    Upload a file to generate a download link that can easily be shared with others. Files that are
                    uploaded stay available to download for 24 hours.
                </h3>
                <a onClick={this.uploadClicked} className="uploadButton" style={{ color: "white" }}>
                    Upload File
                </a>
                <input
                    type="file"
                    id="file"
                    ref="fileUpload"
                    style={{ display: "none" }}
                    onChange={this.onFileSelected}
                />

                {this.state.fileName !== "" ? (
                    <UploadItem
                        fileType={this.state.fileType}
                        fileName={this.state.fileName}
                        expireTime="24 hrs"
                        downloadLink={this.state.downloadLink}
                    />
                ) : (
                    <i />
                )}
                {this.displayProgress()}
            </div>
        );
    }
}
