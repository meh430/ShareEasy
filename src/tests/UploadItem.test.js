import React from "react";
import { shallow } from "enzyme";
import { UploadItem } from "../components/UploadItem";

const mediaIcons = {
    image: "fas fa-file-image",
    video: "far fa-file-video",
    audio: "far fa-file-audio",
    pdf: "far fa-file-pdf",
    text: "far fa-file-alt",
    zip: "far fa-file-archive",
    upload: "fas fa-upload",
};

//fileName: string
//fileType: string
//downloadLink: string
//expireTime: optional string
test("should test all possible Upload Item components in progress", () => {
    for (const type in mediaIcons) {
        const wrapper = shallow(<UploadItem fileName="file name" fileType={type} downloadLink="" />);
        expect(wrapper.instance().displayMediaIcon()).toEqual(mediaIcons[type]);
        expect(wrapper).toMatchSnapshot();
    }
});

test("should test all possible Upload Item components finished", () => {
    for(const type in mediaIcons) {
        const wrapper = shallow(<UploadItem fileName="file name" fileType={type} downloadLink="link" />);
        expect(wrapper.instance().displayMediaIcon()).toEqual(mediaIcons[type]);
        expect(wrapper).toMatchSnapshot();
    }
});


test("should test all no file type Upload Item components", () => {
    const wrapper = shallow(<UploadItem fileName="file name" fileType="" downloadLink="link" />);
    expect(wrapper.instance().displayMediaIcon()).toEqual(mediaIcons.upload);
    expect(wrapper).toMatchSnapshot();
});