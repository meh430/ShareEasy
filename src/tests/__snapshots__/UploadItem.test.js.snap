// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test all no file type Upload Item components 1`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="fas fa-upload"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div
    className="dLink"
    style={
      Object {
        "backgroundColor": "#a5bef0",
      }
    }
  >
    link
     
    <CopyToClipboard
      text="link"
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "borderLeft": "solid 1px black",
            "display": "inline-block",
            "fontSize": "16px",
            "marginLeft": "12px",
            "padding": "8px",
            "paddingLeft": "12px",
          }
        }
      >
        <i
          className="far fa-clone"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        />
      </div>
    </CopyToClipboard>
  </div>
  <div />
</div>
`;

exports[`should test all possible Upload Item components finished 1`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="fas fa-file-image"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div
    className="dLink"
    style={
      Object {
        "backgroundColor": "#a5bef0",
      }
    }
  >
    link
     
    <CopyToClipboard
      text="link"
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "borderLeft": "solid 1px black",
            "display": "inline-block",
            "fontSize": "16px",
            "marginLeft": "12px",
            "padding": "8px",
            "paddingLeft": "12px",
          }
        }
      >
        <i
          className="far fa-clone"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        />
      </div>
    </CopyToClipboard>
  </div>
  <div />
</div>
`;

exports[`should test all possible Upload Item components finished 2`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="far fa-file-video"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div
    className="dLink"
    style={
      Object {
        "backgroundColor": "#a5bef0",
      }
    }
  >
    link
     
    <CopyToClipboard
      text="link"
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "borderLeft": "solid 1px black",
            "display": "inline-block",
            "fontSize": "16px",
            "marginLeft": "12px",
            "padding": "8px",
            "paddingLeft": "12px",
          }
        }
      >
        <i
          className="far fa-clone"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        />
      </div>
    </CopyToClipboard>
  </div>
  <div />
</div>
`;

exports[`should test all possible Upload Item components finished 3`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="far fa-file-audio"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div
    className="dLink"
    style={
      Object {
        "backgroundColor": "#a5bef0",
      }
    }
  >
    link
     
    <CopyToClipboard
      text="link"
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "borderLeft": "solid 1px black",
            "display": "inline-block",
            "fontSize": "16px",
            "marginLeft": "12px",
            "padding": "8px",
            "paddingLeft": "12px",
          }
        }
      >
        <i
          className="far fa-clone"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        />
      </div>
    </CopyToClipboard>
  </div>
  <div />
</div>
`;

exports[`should test all possible Upload Item components finished 4`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="far fa-file-pdf"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div
    className="dLink"
    style={
      Object {
        "backgroundColor": "#a5bef0",
      }
    }
  >
    link
     
    <CopyToClipboard
      text="link"
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "borderLeft": "solid 1px black",
            "display": "inline-block",
            "fontSize": "16px",
            "marginLeft": "12px",
            "padding": "8px",
            "paddingLeft": "12px",
          }
        }
      >
        <i
          className="far fa-clone"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        />
      </div>
    </CopyToClipboard>
  </div>
  <div />
</div>
`;

exports[`should test all possible Upload Item components finished 5`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="far fa-file-alt"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div
    className="dLink"
    style={
      Object {
        "backgroundColor": "#a5bef0",
      }
    }
  >
    link
     
    <CopyToClipboard
      text="link"
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "borderLeft": "solid 1px black",
            "display": "inline-block",
            "fontSize": "16px",
            "marginLeft": "12px",
            "padding": "8px",
            "paddingLeft": "12px",
          }
        }
      >
        <i
          className="far fa-clone"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        />
      </div>
    </CopyToClipboard>
  </div>
  <div />
</div>
`;

exports[`should test all possible Upload Item components finished 6`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="far fa-file-archive"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div
    className="dLink"
    style={
      Object {
        "backgroundColor": "#a5bef0",
      }
    }
  >
    link
     
    <CopyToClipboard
      text="link"
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "borderLeft": "solid 1px black",
            "display": "inline-block",
            "fontSize": "16px",
            "marginLeft": "12px",
            "padding": "8px",
            "paddingLeft": "12px",
          }
        }
      >
        <i
          className="far fa-clone"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        />
      </div>
    </CopyToClipboard>
  </div>
  <div />
</div>
`;

exports[`should test all possible Upload Item components finished 7`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="fas fa-upload"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div
    className="dLink"
    style={
      Object {
        "backgroundColor": "#a5bef0",
      }
    }
  >
    link
     
    <CopyToClipboard
      text="link"
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "borderLeft": "solid 1px black",
            "display": "inline-block",
            "fontSize": "16px",
            "marginLeft": "12px",
            "padding": "8px",
            "paddingLeft": "12px",
          }
        }
      >
        <i
          className="far fa-clone"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        />
      </div>
    </CopyToClipboard>
  </div>
  <div />
</div>
`;

exports[`should test all possible Upload Item components in progress 1`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="fas fa-file-image"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div />
</div>
`;

exports[`should test all possible Upload Item components in progress 2`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="far fa-file-video"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div />
</div>
`;

exports[`should test all possible Upload Item components in progress 3`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="far fa-file-audio"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div />
</div>
`;

exports[`should test all possible Upload Item components in progress 4`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="far fa-file-pdf"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div />
</div>
`;

exports[`should test all possible Upload Item components in progress 5`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="far fa-file-alt"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div />
</div>
`;

exports[`should test all possible Upload Item components in progress 6`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="far fa-file-archive"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div />
</div>
`;

exports[`should test all possible Upload Item components in progress 7`] = `
<div
  className="upItem"
>
  <div
    className="upItemText"
  >
    <i
      className="fas fa-upload"
      style={
        Object {
          "color": "red",
          "fontSize": "52px",
          "margin": "14px",
        }
      }
    />
    <h3
      className="fileName"
    >
      file name
    </h3>
  </div>
  <div />
</div>
`;
